packages/core/src/external/fhir/shared/bundle.ts

```diff
@@ -176,12 +166,44 @@ export function buildBundleFromResources({
   };
 }
 
+export type RequiredBundleType = NonNullable<Bundle["type"]>;
+
+export type BundleType<B extends RequiredBundleType, R extends Resource> = Bundle<R> & {
+  type?: B;
+};
+
+export type ReturnBundleType<B extends RequiredBundleType, R extends Resource> = Required<
+  Pick<BundleType<B, R>, "type">
+> &
+  Omit<BundleType<B, R>, "type">;
+
+export function buildBundle<B extends RequiredBundleType, R extends Resource>({
+  type,
+  entries = [],
+}: {
+  type: B;
+  entries?: BundleEntry<R>[];
+}): ReturnBundleType<B, R> {
+  return {
+    resourceType: "Bundle" as const,
+    total: entries.length,
+    type,
+    entry: entries,
+  };
+}
+
+export function buildCollectionBundle<T extends Resource = Resource>(
+  entries: BundleEntry<T>[] = []
+): CollectionBundle<T> {
+  return buildBundle({ type: "collection", entries });
+}
+
 export function buildSearchSetBundle<T extends Resource = Resource>({
   entries = [],
 }: {
   entries?: BundleEntry<T>[];
 } = {}): SearchSetBundle<T> {
-  return buildBundle({ type: "searchset", entries }) as SearchSetBundle<T>;
+  return buildBundle({ type: "searchset" as const, entries });
```

Comment:
nit: do we need the "as const" here? Just because I see the `buildCollectionBundle` doesn't need it. :)

Reply:
you are correct lol!